lib/std/Build/Watch.zig

```diff
@@ -237,6 +237,300 @@ const Os = switch (builtin.os.tag) {
+                        null,
+                        0,
+                        windows.FILE_SHARE_READ | windows.FILE_SHARE_WRITE | windows.FILE_SHARE_DELETE,
+                        windows.FILE_OPEN,
+                        windows.FILE_DIRECTORY_FILE | windows.FILE_OPEN_FOR_BACKUP_INTENT,
+                        null,
+                        0,
+                    );
+
+                    switch (rc) {
+                        .SUCCESS => {},
+                        .OBJECT_NAME_INVALID => return error.BadPathName,
+                        .OBJECT_NAME_NOT_FOUND => return error.FileNotFound,
+                        .OBJECT_NAME_COLLISION => return error.PathAlreadyExists,
+                        .OBJECT_PATH_NOT_FOUND => return error.FileNotFound,
+                        .NOT_A_DIRECTORY => return error.NotDir,
+                        // This can happen if the directory has 'List folder contents' permission set to 'Deny'
+                        .ACCESS_DENIED => return error.AccessDenied,
+                        .INVALID_PARAMETER => unreachable,
+                        else => return windows.unexpectedStatus(rc),
+                    }
+                }
+                assert(dir_handle != windows.INVALID_HANDLE_VALUE);
+                errdefer windows.CloseHandle(dir_handle);
+
+                const dir_id = try getFileId(dir_handle);
+
+                const dir_ptr = try gpa.create(@This());
+                dir_ptr.* = .{
+                    .handle = dir_handle,
+                    .id = dir_id,
+                    .overlapped = std.mem.zeroes(windows.OVERLAPPED),
+                };
+                return dir_ptr;
+            }
+
+            fn deinit(self: *@This(), gpa: Allocator) void {
+                _ = windows.kernel32.CancelIo(self.handle);
+                windows.CloseHandle(self.handle);
+                gpa.destroy(self);
+            }
+        };
+
+        fn wait(os: Os, timeout: Timeout) !*Directory {
+            var bytes_transferred: windows.DWORD = undefined;
+            var key: usize = undefined;
+            var overlapped_ptr: ?*windows.OVERLAPPED = undefined;
+            while (true) {
+                const status = windows.GetQueuedCompletionStatus(os.io_cp.?, &bytes_transferred, &key, &overlapped_ptr, @bitCast(timeout.to_i32_ms()));
+                switch (status) {
+                    .Normal => {
+                        if (bytes_transferred == 0)
+                            return error.Unexpected;
+                        return os.dir_list.items[key];
+                    },
+                    .Timeout => return error.Timeout,
+                    .Cancelled => continue,
+                    else => return error.Unexpected,
+                }
+            }
+        }
+
+        fn getFileId(handle: windows.HANDLE) !FileId {
+            var file_id: FileId = undefined;
+            {
+                var io_status: windows.IO_STATUS_BLOCK = undefined;
```

Comment:
- It's ok to reuse the iosb since a) we're not doing anything special with it, and b) it shows up in tools like NtTrace - which has been helpful to determine the behaviour of the win32 API.
- Because of the above, you can remove the scoped blocks.
- I would add a trailing comma to the `Nt` calls as it reads better having them broken up over multiple lines. Same goes for the `kernel32` stuff.
- This is just personal taste, but I would also skip assigning an `rc` variable and move the call into the switch like so:
```zig
switch (ntdll.NtQueryFoo(
    &handle,
    ...
)) {
   .SUCCESS => {},
    else => |e| return windows.unexpectedStatus(e),
}
```